app/models/cinstance.rb

```diff
@@ -409,6 +411,10 @@ def change_plan_internal(new_plan)
     end
   end
 
+  def same_service
+    errors.add(:plan_id, :not_allowed) if plan&.service != service
```

Comment:
I don't understand why `plan&.service != service` instead of `plan.service != service` because this is invoked in this case: `validate :same_service, on: :update, if: :plan_id_changed?`
How can `plan_id_changed?` be `true` but at the same time `plan` being `nil` for an `update` 😕

Reply:
The new plan can be `nil`. There's actually a test for that. And in case you're wondering, no, it should not ever be `nil`, but that's covered by other validation: https://github.com/3scale/porta/blob/3be766f22eeb2fb727449e768bf4e2487583769e/app/models/cinstance.rb#L79

Reply:
Ok, I see... before I said that, I tried your test in `test/unit/cinstance_test.rb:680` doing that change that I said and it worked so I was confused.

@guicassolato Can you tell me where is this test that you mentioned?
> There's actually a test for that

I ask because I've seen something interesting. If I run this test that I've just written right below the one you have in `test/unit/cinstance_test.rb:680` right now :
```ruby
test 'checking what happens when updating plan to nil xD' do
  pp @cinstance.change_plan! nil
  pp @cinstance.change_plan nil
  pp @cinstance.valid?
  pp @cinstance.plan
  pp @cinstance.errors.full_messages
end
```
And what I see is no errors of any kind:
![image](https://user-images.githubusercontent.com/11318903/52047645-bde12500-2549-11e9-8c39-42875f12f71b.png)
Even if I change in this line the `plan&.service != service` for `plan.service != service` I see exactly the same 🤔 And I see why in the methods to change plan:
https://github.com/3scale/porta/blob/774f9676967f18dcda839e1c187fb46e029e08cc/app/models/contract.rb#L187-L201
https://github.com/3scale/porta/blob/774f9676967f18dcda839e1c187fb46e029e08cc/app/models/contract.rb#L263-L264
so basically `changed` is `true` when is `nil` although it didn't actually change anything so it just returns the old plan and does not have any error anywhere just like if had done the change... which doesn't look like a good idea to me  😕

Reply:
Mine fails at https://github.com/3scale/porta/blob/3be766f22eeb2fb727449e768bf4e2487583769e/test/unit/cinstance_test.rb#L8

<img width="585" alt="screen shot 2019-01-31 at 3 10 10 pm" src="https://user-images.githubusercontent.com/1842261/52059631-a2d2dd00-256a-11e9-9c47-d0488d489bfd.png">